keps/140-scenario-based-simulation/README.md

```diff
@@ -0,0 +1,397 @@
+  // When PodUnscheduled is non nil, Operation should be "PodUnscheduled".
+  PodUnscheduled  *PodResult
+  // PodPreempted represents the scheduler preempted the Pod.
+  // When PodPreempted is non nil, Operation should be "PodPreempted".
+  PodPreempted  *PodResult
+}
+
+type CreateOperationResult struct {
+  // Object is the Object to be create.
+  Object runtime.Object
+}
+
+type PatchOperationResult struct {
+  TypeMeta   metav1.TypeMeta
+  ObjectMeta metav1.ObjectMeta
+​​  // Patch is the patch for target.
+  Patch string
+  // Result is the resource after patch.
+  Object runtime.Object
+}
+
+type DeleteOperationResult struct {
+  TypeMeta   metav1.TypeMeta
+  ObjectMeta metav1.ObjectMeta
+}
+
+type DoneOperationResult struct {
+  Done bool   
+}
+
+// PodResult has the results related to the specific Pod.
+// Depending on the status of the Pod, some fields may be empty.
+type PodResult struct {
+  Pod v1.Pod
+  // BoundTo indicates to which Node the Pod was scheduled.
+  BoundTo             *string
+  // PreemptedBy indicates which Pod the Pod was deleted for.
+  // This field may be nil if this Pod has not been preempted.
+  PreemptedBy         *string
+  // CreatedAt indicates when the Pod was created.
+  CreatedAt           FakeTime
+  // BoundAt indicates when the Pod was scheduled.
+  // This field may be nil if this Pod has not been scheduled.
+  BoundAt             *FakeTime
+  // PreemptedAt indicates when the Pod was preempted.
+  // This field may be nil if this Pod has not been preempted.
+  PreemptedAt         *FakeTime
+  // ScheduleResult has the results of all scheduling for the Pod.
+  ScheduleResult      []ScenarioPodScheduleResult
+}
+
+type ScenarioPodScheduleResult struct {
+  // Time indicates the time scheduling at which the scheduling is performed.
+  Time                *FakeTime
+  // AllCandidateNodes indicates all candidate Nodes before Filter.
+  AllCandidateNodes   []string
+  // AllFilteredNodes indicates all candidate Nodes after Filter.
+  AllFilteredNodes    []string
+  // PluginResults has each plugin’s result.
+  PluginResults       ScenarioPluginsResults
+}
+
+type (
+  NodeName   string
+  PluginName string
+)
+
+type ScenarioPluginsResults struct {
+  // Filter has each filter plugin’s result.
+  Filter            map[NodeName]map[PluginName]string
+  // Score has each score plugin’s score.
+  Score             map[NodeName]map[PluginName]ScenarioPluginsScoreResult
+}
+
+type ScenarioPluginsScoreResult struct {
+  // RawScore has the score from Score method of Score plugins.
+  RawScore             int64
+  // NormalizedScore has the score calculated by NormalizeScore method of Score plugins.
+  NormalizedScore      int64
+  // FinalScore has score plugin’s final score calculated by normalizing with NormalizedScore and applied Score plugin weight. 
+  FinalScore           int64
+}
+```
+
+#### The concept "FakeTime"
```

Comment:
How about renaming this to `ScenarioStep` or `Step`??  I think `Step` can describe its nature clearly as discrete and sequential. Moreover, `ScenarioPluginsResults` can include timestamps generally because those results depend on their implementation.


Then, we can introduce a section like "Semantics of Simulation Scenario Steps" which can focus only on what happened in the steps.　I think the section makes this easier to read for readers.

Reply:
> How about renaming this to ScenarioStep or Step?? 

+100 💯 
Let me update.